Fix issue #14

jeremymreed opened the pull request:
Catch GLib.Error, and display alternate message when
gnome-pomodoro is not running.

See: https://github.com/kantord/i3-gnome-pomodoro/issues/14

jeremymreed committed:
Fix issue #14

Catch GLib.Error, and display alternate message when
gnome-pomodoro is not running.

See: https://github.com/kantord/i3-gnome-pomodoro/issues/14

jeremymreed commented:
This PR will display "---" if GLib.Error is thrown.  Feel free to change this string to some other value as needed.

kantord commented:
Thanks for the merge request!

Looks good to me! But as you say, maybe we could find a better string?

Unfortunately I don't know how most people use it. I automatically start gnome-pomodoro, so if it's not running that's a really rare occasion. Therefore what I would like to see there personally is something like "gnome-pomodoro is not running".

But if someone only starts it occasionally, maybe they would prefer your version.

Do you have any thoughts on that?

jeremymreed commented:
I'm closing this..  This works to prevent i3-gnome-pomodoro from printing an exception, but breaks other functions in the code.  I'll do a little more work on this PR, then resubmit.  Sorry, was a bit too quick to pull the trigger on this one.

jeremymreed commented:
Hey, I just got back online.
Ahh, that makes sense, you're starting gnome-pomodoro automatically.  I typically turn my pomodoro app off when I'm done using it.
I can set the string to "gnome-pomodoro is not running"
I can also set the other commands (pause, resume, start, stop, skip, reset, toggle, and daemon) to print that error string, and exit if gnome-pomodoro is not running.  Would that work?
I can make those changes, and push up a new pull request.

jeremymreed committed:
Fix commands.  Set better error string.

Modify status, pause, resume, start, stop, skip,
reset, toggle to print an error message if gnome-pomodoro
is not running.

Modify create_workspace_policy() to return True if
gnome_pomodoro is not running.

jeremymreed commented:
Changes are done.
I wasn't sure what the best behavior was for the daemon in the case where gnome-pomodoro isn't running.  I've modified create_workspace_policy() to return True in that case.  I've left the rest of the daemon code alone.

Seems to work well on my machine, let me know if you'd like to see any other changes.

jeremymreed committed:
Merge branch 'master' of https://github.com/kantord/i3-gnome-pomodoro into bugfix/fix-glib-error-when-gnome-pomodoro-isnt-running

kantord commented on pomodoro-client.py:
```diff
@@ -149,6 +182,9 @@ def get_focused_workspace(i3):
 def create_workspace_policy(disabled_during_pomodoro):
     def allowed_workspace(number):
         pomodoro = get_pomodoro_proxy()
+        if pomodoro is None:
```
thanks, I totally forgot about this feature and possibility, as I'm not using it nowadays :smile:

kantord commented on pomodoro-client.py:
```diff
@@ -76,47 +79,77 @@ def main():
     pomodoro = get_pomodoro_proxy()
-    pomodoro_data = extract_pomodoro_data(pomodoro)
-    click.echo(format_output(pomodoro_data))
+    if pomodoro is not None:
```
Your MR looks very good to me, thanks for the contrib :smile:!

Just one note. I can see that this check created a very repetitive pattern in the code.

It always had the line 

```python
pomodoro = get_pomodoro_proxy()
```

but now it has

```python
pomodoro = get_pomodoro_proxy()
    if pomodoro is not None:
        # stuff
    else:
        click.echo("gnome-pomodoro is not running")
```

I am not actively working with Python code anymore, but if I remember correctly, the decorator pattern  is the best way to get rid of this duplication. I imagine something like this:

```python
def command_with_pomodoro_proxy(unwrapped_function):
    @click.command()
    def wrapped_function():
        pomodoro = get_pomodoro_proxy()
        if pomodoro is not None:
            unwrapped_function(pomodoro)
        else:
            click.echo("gnome-pomodoro is not running")
```

then, all the commands could be defined in a simpler way, something like this:


```python
@command_with_pomodoro_proxy
def pause(pomodoro):
    pomodoro.Pause()
```

as it's been a long time since I touched Python code, I'm not 100% confident that this code is correct but I think you get the idea :-)

Anyway, I like the pull request as is, but I'd be really grateful if you could also include that refactor (or something similar, if you have a better suggestion)

jeremymreed replied:
Sure, I can take a look into using decorators here! :)
I'll study them, play around with them a bit, then modify my PR to incorporate decorators.

kantord replied:
Thanks! Feel free to ask for help if you're stuck!

jeremymreed replied:
I've updated the PR to use decorators as you've suggested.
Let me know what you think. :)

kantord replied:
Thanks for updating with decorators. Sorry for being such a difficult reviewer, but I still have a suggestion.

While now you are using decorators and the actual logic is not duplicated, you also introduced a lot of boilerplate, so in the end this decorator implementation didn't solve the redundancy issue.

I see now that there was a problem with my original proposal, as click takes function names for command names, so click cannot be applied inside the decorators.

However, decorators can be chained, so there's no reason why we couldn't apply 2 decorators to the same function, i. e. click doesn't have to be applied inside the decorator.

Therefore, this should be possible:

```python
@click.command()
@with_pomodoro_proxy
def reset(pomodoro_proxy):
    pomodoro_proxy.Reset()
```

That way only 1 extra line is added to each command.

The way to implement that would be something like this:

```python
from functools import wraps

def with_pomodoro_proxy(f):
    @wraps(f)
    def wrapped_command():
        pomodoro = get_pomodoro_proxy()
        if pomodoro is not None:
            f(pomodoro)
        else:
            click.echo("gnome-pomodoro is not running")
    return wrapped_command
```

The reason why `functools.wraps` is required is because manually wrapping will also change the function name. `functools.wraps` will wrap the function with the proxy getter functionality without changing it's name, hence it should work fine with click.

Note that the definition for the inner command takes the proxy as an argument:

```
def reset(pomodoro_proxy):
    pomodoro_proxy.Reset()
```

once you add our new decorator (if it runs at first try  that is), then it should effectively turn it into a function with zero arguments:

```python
@with_pomodoro_proxy
def reset(pomodoro_proxy):
    pomodoro_proxy.Reset()
```

which is because of the signature of the wrapped command inside the decorator:

```
    def wrapped_command():
```


therefore it should now work fine with click with a much easier syntax, only at the cost of making the decorator itself a little bit more complicated

jeremymreed replied:
> Sorry for being such a difficult 

Don't worry about that. :)  Your feedback is very helpful.  I'm not a Python expert, I usually use Python to write quickie scripts for my machine.  So I'm learning quite a bit from your feedback! :)
I'm off to dinner, but I'll take a look at this idea afterwards.  Thanks! :)

kantord commented:
> Ahh, that makes sense, you're starting gnome-pomodoro automatically. I typically turn my pomodoro app off when I'm done using it.

kantord commented:
> Ahh, that makes sense, you're starting gnome-pomodoro automatically. I typically turn my pomodoro app off when I'm done using it.

Yes, I essentially never see the gnome-pomodoro window, unless I want to see the stats for example.

However for someone else who might want to open/close gnome-pomodoro manually (say, to save ram), I'm thinking, the text "gnome-pomodoro is not running" might be too annoying, as it takes up a lot of space. Maybe an empty string would do better after all, but that might be confusing as well

jeremymreed committed:
Create a decorator to get pomodoro instance.

Use decorator function to get the pomodoro instance, and
check it for validity.

If it is valid, call the passed-in function with the pomodoro
instance.

If not, print an error message.

jeremymreed committed:
Implement remaining commands.

Also refactor to make code cleaner, make run_<cmd> functions
inner functions of the <cmd> function they are associated
with.

jeremymreed commented:
I'm taking a look at using nagbar now.

kantord commented:
> I'm taking a look at using nagbar now.

Thanks!!

I suggest you take a look at this line:
https://github.com/kantord/i3-gnome-pomodoro/blob/master/pomodoro-client.py#L141

we already have a function that pops a nagbar, so unlike the decorator stuff, here your job should be easy.

```

show_message("(i3-gnome-pomodoro) Error: gnome-pomodoro is not running", is_error=True)

```

jeremymreed commented:
I've updated my PR to include the nagbar warning.  Code now puts an empty string on the bar, and shows an error message.
There is an issue with the status command though.  If a user decides to configure polybar to exec this script with --nagbar argument, and if gnome-pomodoro is not running, they will get spammed by many nagbar warnings as polybar invokes this script.

jeremymreed commented:
Ahh, thanks for the update, I'll use that error message, and I'll alter the call to show_message to pass true for is_error.

jeremymreed commented:
Perhaps have a different decorator function for the status command with a different behavior?

jeremymreed committed:
Use nagbar to display an error.

If gnome-pomodoro is not running, and the user specifies
the --nagbar option, the code will display a nagbar
with an error message.

jeremymreed commented:
All set.  Code currently requires the user to specifiy --nagbar option if they want nagbar warnings.
I'm wondering if they don't want nagbar warnings should this put a message on the bar?
Or should the nagbar be always used in this case?

kantord commented:
> There is an issue with the status command though. If a user decides to configure polybar to exec this script with --nagbar argument, and if gnome-pomodoro is not running, they will get spammed by many nagbar warnings as polybar invokes this script.

Thanks for testing it with polybar!

That's indeed a real issue.

> Perhaps have a different decorator function for the status command with a different behavior?

Yes, I was also thinking along the same lines.

However, maybe in that special case we shouldn't use a decorator at all. Technically it's possible to use parametric decorators to make our decorator configurable, but the code for that would be really ugly. So I would simply use the decorator for all the commands *except* the status command.

So status would look something like this:

```python
@click.command()
def status():
    pomodoro = get_pomodoro_proxy()
    if pomodoro is None:
        click.echo("")
    pomodoro_data = extract_pomodoro_data(pomodoro)
    click.echo(format_output(pomodoro_data))
```

Technically this is code duplication, but it's not as bad as creating a parametric decorator just for this, or copypasting the decorator code, and it also let's use get rid of these for commands:

```python
 @click.option('--nagbar/--no-nagbar', default=False) 
```

since AFAIK that's currently only used for the daemon and therefore this would change the API of i3-gnome-pomodoro so in other words everyone who is currently using this would have a breaking i3-gnome-pomodoro after update, which we have to avoid if possible.

kantord commented:
> I'm wondering if they don't want nagbar warnings should this put a message on the bar?
Or should the nagbar be always used in this case?

Is this a feature you want personally?

I surely don't need this feature, and I think most users probably don't need it, since the messages would only appear when they accidentally try to interact with gnome-pomodoro and it's not running. And in that case showing the message seems to be reasonable.

If you agree, then I think you only need to implement this feature if you personally want/need it. Otherwise we can wait until someone requests this feature in the issues

If you are looking to experiment a little bit more with the code then of course feel free to implement it. Another alternative that you might consider in this case is a command-line-flag for the "start" command, which would automatically launch gnome-pomodoro if it's not running yet, instead of showing a nagbar.

jeremymreed commented:
No problem!  I use polybar myself, and this script is something I'd like to use for tracking pomodoros.  I'm glad to be able to help.
Ok, I'll take a look at this after dinner, thanks for the feedback. :)

kantord commented:
Closing because the lack of activity. Feel free to reopen
